src/Parsers/ASTWithAlias.h

```diff
@@ -18,6 +18,9 @@ class ASTWithAlias : public IAST
     /// Example: 'WITH pow(2, 2) as a SELECT pow(a, 2)' returns 'pow(a, 2)' instead of 'pow(pow(2, 2), 2)'
     bool prefer_alias_to_column_name = false;
 
+    /// If is true, the AST node is referred by others. This flag doesn't affect AST hash and is only used internally.
+    bool used = false;
```

Comment:
This flag looks a bit foreign for `ASTWithAlias` member. Maybe we can store flags separately? For example build mapping from _ast pointer_ to _bool_ in `QueryNormalizer`, save it in visitor's data, and perform lookup in `removeUnusedAliasesFromWithClause`.

Reply:
Good idea.

Reply:
Hmm, we have to record this `used` flag inside `ASTWithAlias`. Because there are many places that analysis/normalize the query ast multiple times. Further analysis will remove used `with elements` because they are already expanded and are not considered as used anymore.

Reply:
So, we will delete more than actually need or do not delete unused?

Reply:
we will delete more than actually need because we cannot tell whether it's expanded or really unused.